usecase/pokemon.go

```diff
@@ -129,3 +131,81 @@ func (u PokemonUsecase) DeletePokemon(id string) (model.Pokemon, error) {
 
 	return pokemon, err
 }
+
+func (u PokemonUsecase) GetConcurrentPokemonCSV(types string, items string, itemsPerWorker string) (model.PokemonList, error) {
+	var pokemonList model.PokemonList
+	var err error
+
+	itemsInt, _ := strconv.Atoi(items)
+	itemsPerWorkerInt, err := strconv.Atoi(itemsPerWorker)
+	poolSize := utils.GetIntFromDivision(itemsInt, itemsPerWorkerInt)
+	limit := utils.GetIntFromDivision(len(pokemonList), poolSize)
+
+	fmt.Println(poolSize)
+	fmt.Println(limit)
+
+	if err == nil {
+		pokemonList, err := u.c.ReadCSV()
+		var oddPokemonList []model.Pokemon
+		var evenPokemonList []model.Pokemon
+
+		if err != nil {
+			return pokemonList, err
+		}
+
+		chOd := make(chan int)
+		chEv := make(chan int)
+
+		var wg sync.WaitGroup
+		wg.Add(poolSize)
+
+		for i := 0; i < poolSize; i++ {
+			go func(i int) {
+				tmpPokemon := pokemonList[i]
+				if tmpPokemon.Id%2 != 0 {
+					chOd <- tmpPokemon.Id
+				} else if tmpPokemon.Id%2 != 1 {
+					chEv <- tmpPokemon.Id
+				}
+			}(i)
+		}
+
+		go func() {
+			defer close(chOd)
+			for i := range chOd {
+				tmpPokemon := pokemonList[i]
+				fmt.Println("odd", i, ":", tmpPokemon)
+				oddPokemonList = append(oddPokemonList, tmpPokemon)
+				wg.Done()
+			}
+		}()
+
+		go func() {
+			defer close(chEv)
+			for i := range chEv {
```

Comment:
I think you actually need a way to get out of this for loop, if not you may be getting into a deadlock. Maybe after each iteration you could check if a shutdown channel has been closed, if so you can break out of the loop

Reply:
Refactored with the example